### ISSUE TITLE
jsx-filename-extension allow always or as needed

### ISSUE THREAD

**Issue by remcohaszing** [yannickcr/eslint-plugin-react#2742]
`react/jsx-filename-extension` reports files using JSX which aren’t using the `.jsx` extension. However, it doesn’t report files that don’t use JSX syntax, regardless of the extension.

I would like to add an option `allow`, which can be set to `always` or `as-needed`. When this is set to `always` (default), the rule will only report JSX syntax in files not using the `.jsx` extension, as it does now. When this option is set to `as-needed`, this rule will also report files that end with the `.jsx` extension, but don’t use JSX syntax.

**Comment by ljharb** [comment]
That sounds great to me - however, the intention here is that `.js`/`.mjs` should only be used for *standard JS* - meaning, "not using jsx" is not sufficient for a file to be able to properly use those extensions.

In other words, with `as-needed`, I would want the file to only be warned upon if it used no non-ecmascript (stage 4) syntax, and that gets very complex.

**Comment by remcohaszing** [comment]
> however, the intention here is that .js/.mjs should only be used for standard JS - meaning, "not using jsx" is not sufficient for a file to be able to properly use those extensions.

That may be your preference, but that’s not what the rule currently does. I think this is another issue.

I believe adding this option is a good idea regardless of that. Beware it’s opt-in, so the current behaviour stays the same. Personally I use mostly TypeScript nowadays and want to disallow the use of the `.tsx` extension unless it’s necessary.

---

To address the issue you described, another option could be added to require (or disallow in case of `allow: as-needed`) for features other than JSX. I imagine a type selector can be added:

For example:

```yaml
- react/jsx-filename-extension:
    - error
    - selector:
        - DoExpression
        - JSXElement
        - JSXFragment
```

The default selector would be `['JSXElement', 'JSXFragment']`.

An addition to this idea would be to allow presets. I.e.:

```yaml
- react/jsx-filename-extension:
    - error
    - selector:
        - jsx
        - stage-3  # Probably includes stage-2, stage-1, and stage-0.
        - VariableDeclaration  #  Custom selectors are still supported when using presets.
```

The downside of presets is this is a mainenance burden.

Another addition would be to allow esquery selectors if simple node types turn out not to be sufficient.
